d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingState.java

```diff
@@ -0,0 +1,164 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.subsetting;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import java.net.URI;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * State of cluster subsetting
+ */
+public class SubsettingState
+{
+  private static final Logger LOG = LoggerFactory.getLogger(SubsettingState.class);
+  private final Object _lock = new Object();
+  private final SubsettingStrategyFactory _subsettingStrategyFactory;
+
+  /**
+   * Map from serviceName => Partition Id => subset map
+   */
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Map<URI, TrackerClient>>> _weightedSubsetsCache;
+
+  /**
+   * Map from serviceName => Partition Id => all potential clients
+   */
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Set<URI>>> _potentialClientsMap;
+
+  @GuardedBy("_lock")
+  private final Map<String, Integer> _minClusterSubsetSizeMap;
+
+  @GuardedBy("_lock")
+  private long _version;
+
+  @GuardedBy("_lock")
+  private long _peerClusterVersion;
+
+  public SubsettingState(SubsettingStrategyFactory subsettingStrategyFactory)
+  {
+    _version = -1;
+    _peerClusterVersion = -1;
+    _subsettingStrategyFactory = subsettingStrategyFactory;
+    _weightedSubsetsCache = new HashMap<>();
+    _potentialClientsMap = new HashMap<>();
+    _minClusterSubsetSizeMap = new HashMap<>();
+  }
+
+  public long getPeerClusterVersion(String serviceName,
+      int minClusterSubsetSize,
+      int partitionId)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+
+    if (subsettingStrategy == null)
+    {
+      return -1;
+    }
+
+    return subsettingStrategy.getPeerClusterVersion();
+  }
+
+  public Map<URI, TrackerClient> getClientsSubset(String serviceName,
+      int minClusterSubsetSize,
+      int partitionId,
+      Map<URI, TrackerClient> potentialClients,
+      long version)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+
+    if (subsettingStrategy == null)
+    {
+      return potentialClients;
+    }
+
+    long peerClusterVersion = subsettingStrategy.getPeerClusterVersion();
+
+    synchronized (_lock)
```

Comment:
Do we need to share the same lock for all the services, or can we do one lock per service? When we implemented the lock in the strategy, it's one lock per service. My concern is the performance, using one lock may make multiple calls to different services wait on the same lock.

Reply:
Good point. Change to service-based lock in the latest commit